flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,249 @@
+        BasePluginConvention basePluginConvention = getProject().getConvention().getPlugin(BasePluginConvention.class);
+        getDestinationDirectory().set(basePluginConvention.getLibsDir());
+        getArchiveBaseName().convention(getProject().getName());
+        getArchiveExtension().convention("jar");
+        launcherClassName = objects.property(String.class).convention(Flask.Constants.DEFAULT_LAUNCHER_NAME);
+        mainClassName = objects.property(String.class);
+        jvmArgs = objects.listProperty(String.class);
+        javaAgents = objects.domainObjectContainer(JavaAgent.class);
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
+
+        Provider<File> heartbeatJarProvider = getProject().provider(() -> {
+            File dest = new File(FlaskJarTask.this.getTemporaryDir(), HeartbeatAgentResource.instance.getDisplayName());
+            try (OutputStream os = Flask.write(dest, false); InputStream is = HeartbeatAgentResource.instance.read()) {
+                Flask.write2Stream(is, os);
+            }
+            return dest;
+        });
+        includeLibraries(heartbeatJarProvider);
+    }
+
+    @Input
+    public String getLauncherArchiveHash() {
+        return Flask.bytes2Hex(Flask.computeSHA256Digest(LauncherResource.instance::read));
+    }
+
+    @Input
+    public String getHeartbeatAgentHash() {
+        return Flask.bytes2Hex(Flask.computeSHA256Digest(HeartbeatAgentResource.instance::read));
+    }
+
+    @RequiredArgsConstructor
+    private static class StreamAction implements CopyActionProcessingStreamAction {
+
+        private final ZipOutputStream zoos;
+        private final Manifest manifest;
+        private final MessageDigest md;
+        private final boolean archivePreserveFileTimestamps;
+        private final byte[] buffer;
+
+        @Override
+        @SneakyThrows
+        public void processFile(FileCopyDetailsInternal fileCopyDetails) {
+            String entryName = fileCopyDetails.getRelativePath().toString();
+            if (!fileCopyDetails.isDirectory() && entryName.startsWith(Flask.Constants.LIBRARIES_FOLDER)) {
+                Supplier<InputStream> streamSupplier = () -> Flask.read(fileCopyDetails.getFile(), false);
+                Attributes attr = manifest.getEntries().computeIfAbsent(entryName, it -> new Attributes());
+                md.reset();
+                attr.putValue(Flask.ManifestAttributes.ENTRY_HASH,
+                        Base64.getEncoder().encodeToString(Flask.computeDigest(streamSupplier, md)));
+            }
+            if (Flask.Constants.METADATA_FOLDER.equals(entryName)) return;
+            ZipEntry zipEntry = new ZipEntry(entryName);
+
+            if(archivePreserveFileTimestamps) {
+                zipEntry.setLastModifiedTime(FileTime.fromMillis(fileCopyDetails.getLastModified()));
+            } else {
+                zipEntry.setLastModifiedTime(FileTime.fromMillis(0));
+            }
```

Comment:
Heh, Gradle uses "1st February 1980" as its "null" time, for reasons that _should_ be explained in the [`ZipCopyAction`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/file/archive/ZipCopyAction.java) comments. So these lines become:
```java
if (archivePreserveFileTimestamps) {
    zipEntry.setTime(fileCopyDetails.getLastModified());
} else {
    zipEntry.setTime(ZipCopyAction.CONSTANT_TIME_FOR_ZIP_ENTRIES);
}
```
I don't think we need to convert the `long` into a `FileTime` for "last modified" field.